controllers/servers/csi/addons_server.py

```diff
@@ -139,3 +137,22 @@ def ResyncVolume(self, request, context):
 
         logger.info("is replication {} ready: {}".format(replication.name, replication.is_ready))
         return pb2.ResyncVolumeResponse(ready=replication.is_ready)
+
+    @staticmethod
+    def _ensure_replication_idempotency(replication_request, replication, volume):
+        if replication_request.replication_type == array_settings.REPLICATION_TYPE_MIRROR and \
+                replication.copy_type != replication_request.copy_type:
+            message = "replication already exists " \
+                      "but has copy type of {} and not {}".format(replication.copy_type,
+                                                                  replication_request.copy_type)
+            return False, message
+        elif replication.replication_type == array_settings.REPLICATION_TYPE_EAR:
+            if replication.volume_group_id != volume.volume_group_id:
+                message = "replication already exists, " \
+                      "but volume {} belongs to another group {}".format(volume.name, volume.volume_group_name)
+                return False, message
+            if replication.name != replication_request.replication_policy:
+                message = "replication already exists, " \
+                          "but volume {} uses another replication policy {}".format(volume.name, replication.name)
+                return False, message
+        return True, ""
```

Comment:
so you can know if the function returns idempotent by the message returns? just `if message:`..

Reply:
Of course, but it's less elegant from my point of view.

Reply:
its a common logic
you can call it "err" instead of "message" if its helps

Reply:
We have no case of True + message, so I just removed flag and checking only message

Reply:
please name it to error or error_message

Reply:
Done